libyul/AsmParser.cpp

```diff
@@ -192,13 +191,40 @@ std::optional<std::pair<std::string_view, SourceLocation>> Parser::parseSrcComme
 	langutil::SourceLocation const& _commentLocation
 )
 {
-	static std::regex const argsRegex = std::regex(
-		R"~~(^(-1|\d+):(-1|\d+):(-1|\d+)(?:\s+|$))~~"  // index and location, e.g.: 1:234:-1
-		R"~~(("(?:[^"\\]|\\.)*"?)?)~~",                // optional code snippet, e.g.: "string memory s = \"abc\";..."
-		std::regex_constants::ECMAScript | std::regex_constants::optimize
-	);
-	std::match_results<std::string_view::const_iterator> match;
-	if (!regex_search(_arguments.cbegin(), _arguments.cend(), match, argsRegex))
+	CharStream argumentStream (std::string(_arguments), "");
+	Scanner scanner (argumentStream);
+	scanner.setScannerMode(ScannerKind::SourceLocationComment);
+
+	std::string_view tail { _arguments.substr(_arguments.size()) };
+	auto const parseIndex = [](Scanner& _scanner, bool expectTrailingColon) -> std::optional<std::string>
+	{
+		bool negative = false;
+		if (_scanner.currentToken() == Token::Sub)
+		{
+			negative = true;
+			_scanner.next();
+		}
+		if (_scanner.currentToken() != Token::Number)
+			return std::nullopt;
+		if (expectTrailingColon && _scanner.peekNextToken() != Token::Colon)
+			return std::nullopt;
+		if (!isValidDecimal(_scanner.currentLiteral()))
+			return std::nullopt;
+		std::string decimal = (negative ? "-" : "") + _scanner.currentLiteral();
+		_scanner.next();
+		if (expectTrailingColon)
+			_scanner.next();
+		return decimal;
+	};
+	std::optional<std::string> rawSourceIndex = parseIndex(scanner, true);
+	std::optional<std::string> rawStart = parseIndex(scanner, true);
+	std::optional<std::string> rawEnd = parseIndex(scanner, false);
+
+	bool const successfullyScannedIndices = rawSourceIndex && rawStart && rawEnd;
+	size_t const startIndex = static_cast<size_t>(scanner.currentLocation().start);
```

Comment:
In the context of the scanner we call these "positions" rather than "indices". This is especially important in this function since it can easily be confused with `sourceIndex` which represents some number assigned to the source file, so a completely different thing.

That said, I think the more important bit of information here is "what is it the start of?":

```suggestion
	size_t const snippetStart = static_cast<size_t>(scanner.currentLocation().start);
```

Reply:
I'd also rename `successfullyScannedIndices` to something like `locationScannedSuccessfully`. The part it is scanning is the source location. Source index is only the first component of that. The other two are positions in the source.

Similarly, `parseIndex()` would more accurately be called `parseLocationComponent()`.